pkg/auto/release.go

```diff
@@ -0,0 +1,181 @@
+	AssetPath       string
+	ChartVersion    string
+	Chart           string
+	ReleaseYamlPath string
+	ForkRemoteURL   string
+}
+
+// ReleaseVersions represents the versions within the release.yaml file
+type ReleaseVersions map[string][]string
+
+// InitRelease will create the Release struct with access to the necessary dependencies.
+func InitRelease(d *lifecycle.Dependencies, s *lifecycle.Status, v, c, f string) (*Release, error) {
+	r := &Release{
+		git:           d.Git,
+		VR:            d.VR,
+		ChartVersion:  v,
+		Chart:         c,
+		ForkRemoteURL: f,
+	}
+
+	var ok bool
+	var assetVersions []lifecycle.Asset
+
+	assetVersions, ok = s.AssetsToBeReleased[r.Chart]
+	if !ok {
+		assetVersions, ok = s.AssetsToBeForwardPorted[r.Chart]
+		if !ok {
+			return nil, errors.New("no asset version to release for chart:" + r.Chart)
+		}
+	}
+
+	var assetVersion string
+	for _, version := range assetVersions {
+		if version.Version == r.ChartVersion {
+			assetVersion = version.Version
+			break
+		}
+	}
+	if assetVersion == "" {
+		return nil, errors.New("no asset version to release for chart:" + r.Chart + " version:" + r.ChartVersion)
+	}
+
+	r.AssetPath, r.AssetTgz = mountAssetVersionPath(r.Chart, assetVersion)
+
+	// Check again if the asset was already released in the local repository
+	if exist := checkAssetReleased(r.AssetPath); exist {
+		return nil, errors.New("asset already released for chart:" + r.Chart + " version:" + r.ChartVersion)
+	}
+
+	// Check if we have a release.yaml file in the expected path
+	if exist, err := filesystem.PathExists(d.RootFs, path.RepositoryReleaseYaml); err != nil || !exist {
+		return nil, errors.New("release.yaml not found")
+	}
+
+	r.ReleaseYamlPath = filesystem.GetAbsPath(d.RootFs, path.RepositoryReleaseYaml)
+
+	return r, nil
+}
+
+// PullAsset will execute the release porting for a chart in the repository
+func (r *Release) PullAsset() error {
+	if err := r.git.FetchBranch(r.VR.DevBranch); err != nil {
+		return err
+	}
+
+	if err := r.git.CheckFileExists(r.AssetPath, r.VR.DevBranch); err != nil {
+		return fmt.Errorf("asset version not found in dev branch: %w", err)
+	}
+
+	if err := r.git.CheckoutFile(r.VR.DevBranch, r.AssetPath); err != nil {
+		return err
+	}
+
+	if err := r.git.ResetHEAD(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkAssetReleased(chartVersion string) bool {
+	if _, err := os.Stat(chartVersion); err != nil {
+		return false
+	}
+
+	return true
+}
+
+func mountAssetVersionPath(chart, version string) (assetPath string, assetTgz string) {
+	// example: assets/longhorn/longhorn-100.0.0+up0.0.0.tgz
+	assetTgz = chart + "-" + version + ".tgz"
+	assetPath = "assets/" + chart + "/" + assetTgz
+	return
+}
+
+// UpdateReleaseYaml reads and parse the release.yaml file to a struct, appends the new version and writes it back to the file.
+func (r *Release) UpdateReleaseYaml() error {
+	var releaseVersions ReleaseVersions
+
+	rvBytes, err := os.ReadFile(r.ReleaseYamlPath)
```

Comment:
Rather than reading directly into memory, we should use the stream interface methods of encoder and decoder and we can condense the operations down as well as read the stream of bytes. This goes for the unmarshal function below.